text/1444-union.md

```diff
@@ -178,39 +190,489 @@ fn is_zero(v: Value) -> bool {
 }
 ```
 
-Note that a pattern match on a union field that has a smaller size than the
-entire union must not make any assumptions about the value of the union's
-memory outside that field.  For example, if a union contains a `u8` and a
-`u32`, matching on the `u8` may not perform a `u32`-sized comparison over the
-entire union.
+Since union fields share common storage, gaining write access to one
+field of a union can give write access to all its remaining fields.
+Borrow checking rules have to be adjusted to account for this fact.
+As a result, if one field of a union is borrowed, all its remaining fields
+are borrowed as well for the same lifetime.
+
+```rust
+// ERROR: cannot borrow `u` (via `u.f2`) as mutable more than once at a time
+fn test() {
+    let mut u = MyUnion { f1: 1 };
+    unsafe {
+        let b1 = &mut u.f1;
+                      ---- first mutable borrow occurs here (via `u.f1`)
+        let b2 = &mut u.f2;
+                      ^^^^ second mutable borrow occurs here (via `u.f2`)
+        *b1 = 5;
+    }
+    - first borrow ends here
+    assert_eq!(unsafe { u.f1 }, 5);
+}
+```
+
+As you could see, in many aspects (except for layouts, safety and ownership)
+unions behave exactly like structs, largely as a consequence of inheriting their
+syntactic shape from structs.  
+This is also true for many unmentioned aspects of Rust language (such as
+privacy, name resolution, type inference, generics, trait implementations,
+inherent implementations, coherence, pattern checking, etc etc etc).
+
+>Unless some difference is explicitly mentioned in this or the next section,
+behavior of a union should coincide with behavior of an equivalent struct.
+
+# Detailed design
+
+## Definitions
+
+### Union variant
+
+Same as union field. These two terms are further used interchangeably.
+
+### Trivially destructible type
+
+Either of the following:
+
+- primitive scalar type (including references) or `str`, or
```

Comment:
Probably don't need the `, or` at the end of each item here since it's a list. I though there was a markdown rendering error at first.